fact.io.hdureader, IllegalArgumentException

```
java.lang.IllegalArgumentException: The value for THEAP should be equal to (NAXIS1*NAXIS2 + padding to   2880 blocks).
See section 7.3.2 of the FITS 3.0 standard
        at fact.io.hdureader.BinTable.binTableSanityCheck(BinTable.java:256)
        at fact.io.hdureader.BinTable.<init>(BinTable.java:129)
        at fact.io.hdureader.HDU.<init>(HDU.java:92)
        at fact.io.hdureader.FITS.<init>(FITS.java:88)
        at fact.io.hdureader.FITSStream.init(FITSStream.java:41)
        at stream.runtime.ProcessContainer.execute(ProcessContainer.java:619)
        at stream.runtime.ProcessContainer.run(ProcessContainer.java:572)
        at stream.run.main(run.java:171)
        at stream.run.main(run.java:144)
Exception in thread "main" java.lang.RuntimeException: java.lang.IllegalArgumentException: The value for THEAP should be equal to (NAXIS1*NAXIS2 + padding to 2880 blocks).
See section 7.3.2 of the FITS 3.0 standard
        at stream.runtime.ProcessContainer.run(ProcessContainer.java:575)
        at stream.run.main(run.java:171)
        at stream.run.main(run.java:144)
Caused by: java.lang.IllegalArgumentException: The value for THEAP should be equal to (NAXIS1*NAXIS2 +   padding to 2880 blocks).
See section 7.3.2 of the FITS 3.0 standard
        at fact.io.hdureader.BinTable.binTableSanityCheck(BinTable.java:256)
        at fact.io.hdureader.BinTable.<init>(BinTable.java:129)
        at fact.io.hdureader.HDU.<init>(HDU.java:92)
        at fact.io.hdureader.FITS.<init>(FITS.java:88)
        at fact.io.hdureader.FITSStream.init(FITSStream.java:41)
        at stream.runtime.ProcessContainer.execute(ProcessContainer.java:619)
        at stream.runtime.ProcessContainer.run(ProcessContainer.java:572)
        ... 2 more
```
The following runs aborted processing and wrote sth. to stderr. I checked about 10 of them and it was always the error message from above.

20150407_053
20150407_065
20150407_089
20150407_098
20150407_121
20150407_129
20150407_143
20150407_146
20150408_021
20150408_027
20150408_066
20150408_069
20150408_082
20150408_119
20150408_143
20150408_161
20150408_195
20150408_220
20150408_221
20150408_227
20150409_068
20150409_101
20150409_120
20150409_162
20150409_191
20150409_192
20150410_026
20150410_034
20150410_064
20150410_065
20150410_099
20150410_157
20150410_197
20150411_028
20150411_080
20150411_093
20150411_107
20150411_119
20150411_137
20150411_138
20150411_182
20150411_207
20150412_022
20150412_057

The output can be found at ISDC: ``` /gpfs1/scratch/relleums/single_pulse_pass4/```.
In ```std``` is the stdout and stderr of the workernode job script. In ```job```are the worker node job scripts and in ```phs``` is the photon stream.

All errors when processing 1% (random) of all FACT runs.
I made a machine readable JSON collection of all the errors, the exact input files, and exact program calls:
[error_collection.json.txt](https://github.com/fact-project/fact-tools/files/738867/error_collection.json.txt)

--- a/src/main/java/fact/io/hdureader/BinTable.java
+++ b/src/main/java/fact/io/hdureader/BinTable.java
@@ -182,6 +182,17 @@ public class BinTable {
             Integer naxis1 = header.getInt("NAXIS1").orElse(0);
             Integer naxis2 = header.getInt("NAXIS2").orElse(0);
 
+            // In ZFITS file the heap data starts at the offset stored in the 'ZHEAPPTR' variable in the header.
+            // This is of course documented nowhere. Not even in the 'official spec' https://arxiv.org/pdf/1506.06045v1.pdf
+            // Its also simply wrong according to the FITS standard. It should be the 'THEAP' variable.
+            // However in this case we give more priority to the mysterious ZHEAPPTR variable if its there.
+            // Heres an email I got from Etienne concerning this keyword:
+            //
+            //      I remember that there is a mismatch between the THEAP (or ZHEAPPTR) values that we output to our .fits.fz
+            //      files and the FITS standard. At the time, it was because the FTOOLS had a bug and I had implemented the
+            //      version that worked with the FTOOLS rather than the version compliant with the standard.
+            //
+            //
             int theap = header.getInt("ZHEAPPTR").orElseGet(() -> header.getInt("THEAP").orElse(naxis1*naxis2));
 
             int bytesToSkip = Math.toIntExact((hduOffset + header.headerSizeInBytes + theap));
@@ -238,23 +249,26 @@ public class BinTable {
         if(pcount < 0){
             throw new IllegalArgumentException("Number of bytes in Heap (PCOUNT) is negative.");
         }
-
+        /*
+            See page 25 of the Fits standard:
+                The zero indexed byte offset to the start of
+                the heap, measured from the start of the main data table, may
+                be given by the THEAP keyword in the header
+                ...
+                (i.e., the default value of THEAP is NAXIS1 × NAXIS2)
+                This default value is the minimum allowed
+                value for the THEAP keyword, because any smaller value would
+                imply that the heap and the main data table overlap.
+         */
         if (pcount > 0){
-            int theap = header.getInt("THEAP").orElseThrow(() -> {
-                log.error("The THEAP keyword cannot be found in the BinTable. " +
-                        "Its mandatory when PCOUNT is larger than 0.\nSee section 7.3.2 of the FITS 3.0 standard");
-                return new IllegalArgumentException("Missing THEAP keyword");
-
-            });
+            int theap = header.getInt("THEAP").orElse(naxis1 * naxis2);
 
             if(theap < 0){
                 throw new IllegalArgumentException("Number of bytes to skip to the heap (THEAP) is negative.");
             }
 
-            int desiredTheap = naxis1*naxis2 + (naxis1 * naxis2) % 2880;
-            if(theap == desiredTheap){
-                throw new IllegalArgumentException("The value for THEAP should be equal to (NAXIS1*NAXIS2 + padding to 2880 blocks)." +
-                        "\nSee section 7.3.2 of the FITS 3.0 standard");
+            if(theap < naxis1*naxis2){
+                throw new IllegalArgumentException("The value of THEAP is lower than the allowed minimum of (NAXIS1 * NAXIS2).");
             }
         }
     }

